contracts/Curation.sol

```diff
@@ -0,0 +1,415 @@
+    uint256 constant public PCT_BASE = 10 ** 18; // 0% = 0; 1% = 10^16; 100% = 10^18
+    bytes32 constant public MIN_DEPOSIT_PARAM = keccak256("MIN_DEPOSIT_PARAM");
+    bytes32 constant public APPLY_STAGE_LEN_PARAM = keccak256("APPLY_STAGE_LEN_PARAM");
+    bytes32 constant public DISPENSATION_PCT_PARAM = keccak256("DISPENSATION_PCT_PARAM");
+
+    IRegistry public registry;
+    IStaking public staking;
+    IVoting public voting;
+    uint256 public minDeposit;
+    uint64 public applyStageLen;
+    uint256 public dispensationPct;
+
+    bytes32 constant public CHANGE_PARAMS_ROLE = keccak256("CHANGE_PARAMS_ROLE");
+    bytes32 constant public CHANGE_VOTING_APP_ROLE = keccak256("CHANGE_VOTING_APP_ROLE");
+
+    // TODO!!!
+    enum TimeUnit { Blocks, Seconds }
+
+    struct Application {
+        address applicant;
+        uint64 date;
+        bool registered;
+        bytes data;
+        uint256 amount;
+        uint256 lockId;
+    }
+
+    struct Challenge {
+        address challenger;
+        uint64 date;
+        bool resolved;
+        uint256 amount;
+        uint256 lockId;
+        uint256 voteId;
+        uint256 dispensationPct;
+        mapping(address => bool) claims; // participants who already claimed their reward
+    }
+
+    mapping(bytes32 => Application) applications;
+    mapping(bytes32 => Challenge) challenges;
+    mapping(address => mapping(uint256 => bool)) usedLocks;
+
+    event NewApplication(bytes32 entryId, address applicant);
+    event NewChallenge(bytes32 entryId, address challenger);
+    event ResolvedChallenge(bytes32 entryId, bool result);
+
+    /**
+     * @notice Initializes Curation app with
+     * @param _registry TODO
+     */
+    function initialize(
+        IRegistry _registry,
+        IStaking _staking,
+        IVoting _voting,
+        uint256 _minDeposit,
+        uint64 _applyStageLen,
+        uint256 _dispensationPct
+    )
+        onlyInit
+        external
+    {
+        initialized();
+
+        require(isContract(_registry));
+        require(isContract(_staking));
+
+        registry = _registry;
+        staking = _staking;
+        _setVotingApp(_voting);
+
+        _setMinDeposit(_minDeposit);
+        _setApplyStageLen(_applyStageLen);
+        _setDispensationPct(_dispensationPct);
+    }
+
+    function newApplication(bytes data, uint256 lockId) isInitialized public returns (bytes32 entryId) {
+        entryId = keccak256(data);
+
+        require(data.length != 0);
+        // check data doesn't have an ongoing application
+        require(!applicationExists(entryId));
+        // check data doesn't exist in Registry
+        require(!registry.exists(entryId));
+
+        // check locked tokens
+        uint256 amount = _checkLock(msg.sender, lockId, MAX_UINT64);
+
+        applications[entryId] = Application({
+            applicant: msg.sender,
+            date: getTimestamp(),
+            registered: false,
+            data: data,
+            amount: amount,
```

Comment:
We should enforce that at least the application amount is at least the `minDeposit` in order to apply to prevent spam

Reply:
Oh, I totally missed it

Reply:
😅